In [1]:
import os
import pandas as pd

In [2]:
main_task_path = os.path.join('data', 'main_task.csv')

In [3]:
df = pd.read_csv(main_task_path)

In [4]:
df.head()

Restaurant_id       City                                      Cuisine Style  \
0       id_5569      Paris            ['European', 'French', 'International']   
1       id_1535  Stockholm                                                NaN   
2        id_352     London  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...   
3       id_3456     Berlin                                                NaN   
4        id_615     Munich  ['German', 'Central European', 'Vegetarian Fri...   

   Ranking  Rating Price Range  Number of Reviews  \
0   5570.0     3.5    $$ - $$$              194.0   
1   1537.0     4.0         NaN               10.0   
2    353.0     4.5        $$$$              688.0   
3   3458.0     5.0         NaN                3.0   
4    621.0     4.0    $$ - $$$               84.0   

                                             Reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   
3                                           [[], []]   
4  [['Best place to try a Bavarian food', 'Nice b...   

                                              URL_TA     ID_TA  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781  
3  /Restaurant_Review-g187323-d1358776-Reviews-Es...  d1358776  
4  /Restaurant_Review-g187309-d6864963-Reviews-Au...  d6864963

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  object 
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        26114 non-null  object 
 6   Number of Reviews  37457 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), object(7)
memory usage: 3.1+ MB


In [6]:
# Заполним пропуски в столбце количества отзывов
df['Number of Reviews'].fillna(0, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  object 
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        26114 non-null  object 
 6   Number of Reviews  40000 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), object(7)
memory usage: 3.1+ MB


In [7]:
# Превратим данные в столбце Restaurant_id в числовые данные
df['Restaurant_id'] = df['Restaurant_id'].apply(lambda id: int(id[3:]))

In [8]:
df.head()

Restaurant_id       City  \
0           5569      Paris   
1           1535  Stockholm   
2            352     London   
3           3456     Berlin   
4            615     Munich   

                                       Cuisine Style  Ranking  Rating  \
0            ['European', 'French', 'International']   5570.0     3.5   
1                                                NaN   1537.0     4.0   
2  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...    353.0     4.5   
3                                                NaN   3458.0     5.0   
4  ['German', 'Central European', 'Vegetarian Fri...    621.0     4.0   

  Price Range  Number of Reviews  \
0    $$ - $$$              194.0   
1         NaN               10.0   
2        $$$$              688.0   
3         NaN                3.0   
4    $$ - $$$               84.0   

                                             Reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   
3                                           [[], []]   
4  [['Best place to try a Bavarian food', 'Nice b...   

                                              URL_TA     ID_TA  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781  
3  /Restaurant_Review-g187323-d1358776-Reviews-Es...  d1358776  
4  /Restaurant_Review-g187309-d6864963-Reviews-Au...  d6864963

### 1.3.1 Вопросы о ценах

#### Сколько вариантов непустых значений встречается в столбце 'Price Range'?

In [9]:
price_ranges = pd.Series(df[~df['Price Range'].isna()]['Price Range'])
display(price_ranges.unique())

array(['$$ - $$$', '$$$$', '$'], dtype=object)

#### Ответ: 3

#### Как в датафрейме обозначается самый низкий уровень цен?

#### Ответ: можно догадаться, что '$'

#### Как в датафрейме обозначается самый высокий уровень цен?

#### Ответ: можно догадаться, что \\$\\$\\$\\$

#### Сколько ресторанов относятся к среднему ценовому сегменту?

In [10]:
display(price_ranges.value_counts()[0])

18412

#### Ответ: 18412

In [11]:
# Заполним недостающие данные минимальным уровнем цен
df['Price Range'].fillna('$', inplace=True)
# И проверим изменение картины недостающих данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  int64  
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        40000 non-null  object 
 6   Number of Reviews  40000 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), int64(1), object(6)
memory usage: 3.1+ MB


In [12]:
# На основе этих фактов создадим столбец с числовыми данными об уровне цен так, чтобы:
# $ соответствовал 1, $$-$$$ - 100, а $$$$ - 1000
def get_price_level(price_range):
    if type(price_range) == str:
        if price_range == '$':
            return 1
        if price_range == '$$-$$$':
            return 100
        if price_range == '$$$$':
            return 1000
    return 100

df['price_level'] = df['Price Range'].apply(get_price_level)

### 1.3.2 Вопросы о городах 

#### Сколько городов представлено в наборе данных?

In [13]:
df['City'].nunique()

31

#### Ответ: 31

#### Сколько ресторанов расположены в столичных городах?

In [14]:
# Формируем множество столиц из датасета, найденного на data.world

In [15]:
import json as js

In [16]:
with open('data/country-by-capital-city.json') as json_file:
    country_capital_city_dict = js.load(json_file)

In [17]:
capitals = set()
for row in country_capital_city_dict:
    capitals.add(row['city'])

In [18]:
# Добавим функцию определяющую входит ли город в множество столиц
def is_capital(city):
    return city in capitals

# Применим эту функцию к столбцу 'City'и суммируем результат
df[df['City'].apply(is_capital)]['Restaurant_id'].nunique()

11242

In [19]:
df[df['City'].apply(is_capital)]['City']

0             Paris
1         Stockholm
2            London
3            Berlin
7        Bratislava
            ...    
39992     Stockholm
39993         Paris
39994         Paris
39996         Paris
39997     Stockholm
Name: City, Length: 21436, dtype: object

In [20]:
df.groupby('Restaurant_id')['Restaurant_id'].count().head(20)

Restaurant_id
0     15
1      9
2     14
3      8
4     13
5     11
6      8
7     11
8     13
9     12
10    14
11     8
12    14
13    11
14    15
15     7
16     8
17    14
18    10
19    13
Name: Restaurant_id, dtype: int64

In [21]:
df[df['Restaurant_id'] == 0]['City']

4453       Hamburg
5441          Rome
5936        London
6194         Milan
7199        Madrid
12734         Oslo
12803    Stockholm
19744       Krakow
21372         Lyon
21805       Lisbon
22863    Edinburgh
23201       Vienna
23892       Warsaw
24120    Amsterdam
34458     Budapest
Name: City, dtype: object

In [22]:
sum(df['Restaurant_id'] == 0)

15

#### Ответ: 21436 - неправильный!!! Причины: был выбран недостаточно информативный датасет по городам. В этом датасете присутствовали только первичные столицы, а административные столицы отсутствовали. Вывод: использовать правильный датасет.

#### Благодаря члену группы sfdatascience1.slack.com с ником Mikhail решение проблемы было найдено в использовании датасета https://www.kaggle.com/viswanathanc/world-cities-datasets#worldcities.csv и в расширении понятия 'столица' кроме тех городов в поле 'capital', которых стоит значение 'primary' еще и городом 'Edinburgh'. Весьма странно, но куда деваться??? Еще раз огромное спасибо Mikhail за титаническую работу, которую он проделал.

In [23]:
cities = pd.read_csv('data/worldcities.csv')

In [24]:
cities[cities['city'] == 'Edinburgh']

city city_ascii      lat     lng         country iso2 iso3  \
1036  Edinburgh  Edinburgh  55.9483 -3.2191  United Kingdom   GB  GBR   

              admin_name capital  population          id  
1036  Edinburgh, City of   admin    504966.0  1826492520

In [25]:
cities.head()

city   city_ascii      lat       lng        country iso2 iso3  \
0        Tokyo        Tokyo  35.6850  139.7514          Japan   JP  JPN   
1     New York     New York  40.6943  -73.9249  United States   US  USA   
2  Mexico City  Mexico City  19.4424  -99.1310         Mexico   MX  MEX   
3       Mumbai       Mumbai  19.0170   72.8570          India   IN  IND   
4    São Paulo    Sao Paulo -23.5587  -46.6250         Brazil   BR  BRA   

         admin_name  capital  population          id  
0             Tōkyō  primary  35676000.0  1392685764  
1          New York      NaN  19354922.0  1840034016  
2  Ciudad de México  primary  19028000.0  1484247881  
3       Mahārāshtra    admin  18978000.0  1356226629  
4         São Paulo    admin  18845000.0  1076532519

In [26]:
cities['capital'].unique()

array(['primary', nan, 'admin', 'minor'], dtype=object)

#### Как видно в датафрейме cities не столицами считаются только те города, поле capital которых не заполнено никакими данными

In [27]:
# Вычистим города с отсутствующими данными в поле 'capital'
capitals_df = cities.dropna(subset=['capital'])

# Сформируем множество столиц всех типов, кроме типа 'minor'
capitals = set(capitals_df[capitals_df['capital'] == 'primary']['city'])

In [28]:
capitals.add('Edinburgh')

In [29]:
# Применим эту функцию к столбцу 'City'и суммируем результат
sum(df['City'].apply(is_capital))

30424

#### Ответ: 30424, блин!!!

In [30]:
# А теперь на основе этих фактов создадим числовой столбец is_capital
df['is_capital'] = df['City'].apply(is_capital)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  int64  
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        40000 non-null  object 
 6   Number of Reviews  40000 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
 10  price_level        40000 non-null  int64  
 11  is_capital         40000 non-null  bool   
dtypes: bool(1), float64(3), int64(2), object(6)
memory usage: 3.4+ MB


###  1.3.3. Вопросы о кухнях

#### Сколько типов кухонь представлено в наборе данных?

In [31]:
# Для начала получим Series столбца 'Cuisine Style' без пропусков данных
cuisines = df['Cuisine Style'].dropna()
cuisines

0                  ['European', 'French', 'International']
2        ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...
4        ['German', 'Central European', 'Vegetarian Fri...
6                                     ['Italian', 'Pizza']
7                                              ['Italian']
                               ...                        
39995    ['Italian', 'Vegetarian Friendly', 'Vegan Opti...
39996    ['French', 'American', 'Bar', 'European', 'Veg...
39997                                ['Japanese', 'Sushi']
39998    ['Polish', 'European', 'Eastern European', 'Ce...
39999                                          ['Spanish']
Name: Cuisine Style, Length: 30717, dtype: object

In [32]:
type(cuisines.iloc[0])

str

#### Как видим, данные в этом Series представляют собой строковые представления списков в соответствии с синтаксисом языка Python.

In [33]:
# Применим библиотеку ast для перевода строковых представлений списков в соответствии с ситнтаксисом языка Python в списки Python
import ast

def get_list(list_string):
    result_list = ast.literal_eval(list_string)
    return result_list

# Преобразуем данные в столбце 'Cuisine Style' к списку
cuisines = cuisines.apply(get_list)

# Создадим словарь кухонь
cuisines_dict = dict()

for cuisines_list in cuisines:
    for cuisine in cuisines_list:
        try:
            cuisines_dict[cuisine] += 1
        except:
            cuisines_dict[cuisine] = 1

# Выведем количество различных кухонь

print(f'Множество различных кухонь: {len(cuisines_dict)}')

Множество различных кухонь: 125


#### Какая кухня представлена в наибольшем количестве ресторанов? Введите название кухни без кавычек или апострофов.

In [34]:
sorted(cuisines_dict.items(), key = lambda kv:(kv[1], kv[0]), reverse=True)[0]

('Vegetarian Friendly', 11189)

#### Сколько типов кухонь предлагается только в одном ресторане? Введите ответ в виде целого числа.

In [35]:
uno_cuisine_count = 0
unique_cuisines = set()
for cuisine, count in cuisines_dict.items():
    if count == 1:
        unique_cuisines.add(cuisine)
        uno_cuisine_count += 1
print(f'Количество типов кухонь, предлагаемых только в одном ресторане: {uno_cuisine_count}')
print(f'Уникальные кухни:')
display(unique_cuisines)

Количество типов кухонь, предлагаемых только в одном ресторане: 5
Уникальные кухни:


{'Burmese', 'Latvian', 'Salvadoran', 'Xinjiang', 'Yunnan'}

#### Какое среднее количество кухонь предлагается в одном ресторане? Если в данных отсутствует информация о типах кухонь, то считайте, что в этом ресторане предлагается только один тип кухни. Ответ округлите до одного знака после запятой.

In [36]:
def get_cuisines_count(cuisines):
    if type(cuisines) == str:
        return len(get_list(cuisines))
    return 1

In [37]:
# Заполним недостающие данные в столбце типов кухонь
df['Cuisine Style'].fillna("['usual cuisine']", inplace=True)

In [38]:
# На основании этих фактов создадим числовой столбец 'cuisines_count'
df['cuisines_count'] = df['Cuisine Style'].apply(get_cuisines_count)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  int64  
 1   City               40000 non-null  object 
 2   Cuisine Style      40000 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        40000 non-null  object 
 6   Number of Reviews  40000 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
 10  price_level        40000 non-null  int64  
 11  is_capital         40000 non-null  bool   
 12  cuisines_count     40000 non-null  int64  
dtypes: bool(1), float64(3), int64(3), object(6)
memory usage: 3.7+ MB


In [39]:
# На основе этих данных создадим столбец с числовыми данными is_unique_cuisine, указывающими на уникальную кухню данного ресторана
def is_unique_cuisine(cuisine):
    cuisines_list = get_list(cuisine)
    cuisines_set = set(cuisines_list)
    return not cuisines_set.isdisjoint(unique_cuisines)

df['unique_cuisine'] = df['Cuisine Style'].apply(is_unique_cuisine)

In [40]:
mean_cuisines_count = df['cuisines_count'].mean()
print(f'Среднее количество кухонь предлагаемых в одном ресторане: {mean_cuisines_count:.1f}')

Среднее количество кухонь предлагаемых в одном ресторане: 2.6


### 1.3.4 Вопросы об отзывах

#### Когда был оставлен самый свежий отзыв? Введите ответ в формате yyyy-mm-dd.

In [41]:
# Сначала заменим nan в строковых представлениях списков отзывов и их дат на строку с датой по-умолчанию
# а также заменим пустой список вида [] на список с незаполненными элементами вида ['01/01/2000', '01/01/2000']
def nan_to_default_date(list_string):
    list_string = list_string.replace('[nan', "['01/01/2000'")
    list_string = list_string.replace('nan]', "'01/01/2000']")
    list_string = list_string.replace('[]', "['01/01/2000', '01/01/2000']")
    return list_string

df['Reviews'] = df['Reviews'].apply(nan_to_default_date)    

In [42]:
# Преобразуем данные в столбце 'Reviews' к типу list
last_review = []
last_review_date = []
prelast_review = []
prelast_review_date = []
for reviews in df['Reviews']:
    reviews_list = get_list(reviews)
    if len(reviews_list) == 2:
        if (len(reviews_list[0]) == 2 and len(reviews_list[1]) == 2):
            last_review.append(reviews_list[0][0])
            last_review_date.append(reviews_list[1][0])
            prelast_review.append(reviews_list[0][1])
            prelast_review_date.append(reviews_list[1][1])
        elif (len(reviews_list[0]) == 1 and len(reviews_list[1]) == 1):
            last_review.append(reviews_list[0][0])
            last_review_date.append(reviews_list[1][0])
            prelast_review.append('None review')
            prelast_review_date.append('01/01/2000')
        else:
            print(reviews_list)

In [43]:
df['last_review'] = last_review
df['last_review_date'] = last_review_date
df['prelast_review'] = prelast_review
df['prelast_review_date'] = prelast_review_date

In [44]:
df['last_review_date'] = pd.to_datetime(df['last_review_date'])
df['prelast_review_date'] = pd.to_datetime(df['prelast_review_date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Restaurant_id        40000 non-null  int64         
 1   City                 40000 non-null  object        
 2   Cuisine Style        40000 non-null  object        
 3   Ranking              40000 non-null  float64       
 4   Rating               40000 non-null  float64       
 5   Price Range          40000 non-null  object        
 6   Number of Reviews    40000 non-null  float64       
 7   Reviews              40000 non-null  object        
 8   URL_TA               40000 non-null  object        
 9   ID_TA                40000 non-null  object        
 10  price_level          40000 non-null  int64         
 11  is_capital           40000 non-null  bool          
 12  cuisines_count       40000 non-null  int64         
 13  unique_cuisine       40000 non-

In [45]:
newest_review_date = str(df['last_review_date'].dropna().sort_values(ascending=False).iloc[0])[:10]
print(f'Дата самого свежего отзыва: {newest_review_date}')

Дата самого свежего отзыва: 2018-02-26


#### Какое максимальное количество дней отделяет даты публикации отзывов, размещённых на сайте ресторана? Введите количество дней в виде целого числа.

In [46]:
clear_df = df.dropna(subset=['last_review_date', 'prelast_review_date'])
clear_df['timedelta'] = (df['last_review_date'] - df['prelast_review_date'])
def get_days(timedelta):
    return timedelta.days
clear_df['timedelta'] = clear_df['timedelta'].apply(get_days)
max_days_between_reviews = clear_df['timedelta'].sort_values(ascending=False).iloc[0]
print(f'Максимальное количество дней между датами публикации отзывов, размещенных на сайте ресторана составило {max_days_between_reviews} дней.')

Максимальное количество дней между датами публикации отзывов, размещенных на сайте ресторана составило 6628 дней.


In [47]:
clear_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40000 entries, 0 to 39999
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Restaurant_id        40000 non-null  int64         
 1   City                 40000 non-null  object        
 2   Cuisine Style        40000 non-null  object        
 3   Ranking              40000 non-null  float64       
 4   Rating               40000 non-null  float64       
 5   Price Range          40000 non-null  object        
 6   Number of Reviews    40000 non-null  float64       
 7   Reviews              40000 non-null  object        
 8   URL_TA               40000 non-null  object        
 9   ID_TA                40000 non-null  object        
 10  price_level          40000 non-null  int64         
 11  is_capital           40000 non-null  bool          
 12  cuisines_count       40000 non-null  int64         
 13  unique_cuisine       40000 non-

In [48]:
# Удалим все столбцы, которые содержат нечисловые данные
non_numeric_fields = [
    'Restaurant_id',
    'City',
    'Cuisine Style',
    'Price Range',
    'Reviews',
    'URL_TA',
    'ID_TA',
    'last_review',
    'last_review_date',
    'prelast_review',
    'prelast_review_date'
]
clear_df = clear_df.drop(non_numeric_fields, axis=1)

In [49]:
# Посмотрим на таблицу
clear_df.head()

Ranking  Rating  Number of Reviews  price_level  is_capital  \
0   5570.0     3.5              194.0          100        True   
1   1537.0     4.0               10.0            1        True   
2    353.0     4.5              688.0         1000        True   
3   3458.0     5.0                3.0            1        True   
4    621.0     4.0               84.0          100       False   

   cuisines_count  unique_cuisine  timedelta  
0               3           False         41  
1               1           False        382  
2               7           False          2  
3               1           False          0  
4               3           False        272

In [50]:
# Проверим таблицу на предмет отсутствующих данных
clear_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40000 entries, 0 to 39999
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Ranking            40000 non-null  float64
 1   Rating             40000 non-null  float64
 2   Number of Reviews  40000 non-null  float64
 3   price_level        40000 non-null  int64  
 4   is_capital         40000 non-null  bool   
 5   cuisines_count     40000 non-null  int64  
 6   unique_cuisine     40000 non-null  bool   
 7   timedelta          40000 non-null  int64  
dtypes: bool(2), float64(3), int64(3)
memory usage: 2.2 MB


In [51]:
# Как видно, данные присутствуют везде

# Разбиваем датафрейм на части, необходимые для обучения и тестирования модели

In [52]:
# Х - данные с информацией о ресторанах, у - целевая переменная (рейтинги ресторанов)
X = clear_df.drop(['Rating'], axis = 1)
y = clear_df['Rating']

In [53]:
# Загружаем специальный инструмент для разбивки:
from sklearn.model_selection import train_test_split

In [54]:
# Наборы данных с меткой "train" будут использоваться для обучения модели, "test" - для тестирования.
# Для тестирования мы будем использовать 25% от исходного датасета.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=101)

# Создаём, обучаем и тестируем модель

In [55]:
# Импортируем необходимые библиотеки:
from sklearn.ensemble import RandomForestRegressor # инструмент для создания и обучения модели
from sklearn import metrics # инструменты для оценки точности модели

In [56]:
# Создаём модель
regr = RandomForestRegressor(n_estimators=100)

# Обучаем модель на тестовом наборе данных
regr.fit(X_train, y_train)

# Используем обученную модель для предсказания рейтинга ресторанов в тестовой выборке.
# Предсказанные значения записываем в переменную y_pred
y_pred = regr.predict(X_test)

In [57]:
# Сравниваем предсказанные значения (y_pred) с реальными (y_test), и смотрим насколько они в среднем отличаются
# Метрика называется Mean Absolute Error (MAE) и показывает среднее отклонение предсказанных значений от фактических.
print('MAE:', metrics.mean_absolute_error(y_test, y_pred))

MAE: 0.35470230595238095
